Scape config `job_name` is used without sanitisation for metric name, causing panic

<!-- Please keep the structure below, or your issue may be closed. -->

**Describe the bug**
When attempting to run promtail docker with loki_push_api configured as a scrape_config the docker crashes with the following errors

goroutine 1 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000709860, {0xc0007098a0, 0x1, 0xc0000ee918})
 /src/loki/vendor/github.com/prometheus/client_golang/prometheus/registry.go:403 +0x7f
github.com/prometheus/client_golang/prometheus.MustRegister(...)
 /src/loki/vendor/github.com/prometheus/client_golang/prometheus/registry.go:178
github.com/weaveworks/common/server.New({{0xc0000ee918, 0x14}, {0x300c480, 0x3}, {0x0, 0x0}, 0xdac, 0x0, {0x300c480, 0x3}, ...})
 /src/loki/vendor/github.com/weaveworks/common/server/server.go:172 +0x20d
github.com/grafana/loki/clients/pkg/promtail/targets/lokipush.(*PushTarget).run(0xc0007dfec0)
 /src/loki/clients/pkg/promtail/targets/lokipush/pushtarget.go:96 +0x2b8
github.com/grafana/loki/clients/pkg/promtail/targets/lokipush.NewPushTarget({0x37cedc0, 0xc000108690}, {0x3802bf0, 0xc0007096a0}, {0x0, 0x0, 0x0}, {0xc0006a07d0, 0xb}, 0xc00081ec80)
 /src/loki/clients/pkg/promtail/targets/lokipush/pushtarget.go:76 +0x3f4
github.com/grafana/loki/clients/pkg/promtail/targets/lokipush.NewPushTargetManager({0x3805b20, 0xc000108aa0}, {0x37cedc0, 0xc000108690}, {0x7f52505e8048, 0xc0004b5fc0}, {0xc0007346c0, 0x445460, 0x1})
 /src/loki/clients/pkg/promtail/targets/lokipush/pushtargetmanager.go:47 +0x2ea
github.com/grafana/loki/clients/pkg/promtail/targets.NewTargetManagers({0x37cfc40, 0xc0007df080}, {0x3805b20, 0xc000108aa0}, {0x37cedc0, 0xc000108690}, {0x2540be400, {0xc0000ee7c8, 0x13}, 0x0, ...}, ...)
 /src/loki/clients/pkg/promtail/targets/manager.go:207 +0xf18
github.com/grafana/loki/clients/pkg/promtail.New({{{{0x0, 0x0}, {0x300c480, 0x3}, {0x0, 0x0}, 0x2378, 0x0, {0x300c480, 0x3}, ...}, ...}, ...}, ...)
 /src/loki/clients/pkg/promtail/promtail.go:84 +0x43e
main.main()
 /src/loki/clients/cmd/promtail/main.go:115 +0x61d
level=info ts=2022-09-28T07:15:52.266006699Z caller=pushtarget.go:85 msg="starting push server" job=central-log
panic: descriptor Desc{fqName: "promtail_central-log_tcp_connections", help: "Current number of accepted TCP connections.", constLabels: {}, variableLabels: [protocol]} is invalid: "promtail_central-log_tcp_connections" is not a valid metric name

Removing the loki_push_api scrape config section, promtail starts fine and pushes the other scrape config logs to loki.

I have also attempted to separate the promtails (1 for host logs and 1 for central) to confirm and issue still occurs.

**To Reproduce**
Steps to reproduce the behavior:
1. Start promtail docker with the following config file

server:
  http_listen_port: 9080
  grpc_listen_port: 0
        
positions:
  filename: /tmp/positions.yaml
        
clients:
  - url: https://xxxxx/loki/api/v1/push
    basic_auth: 
      username: xxxx
      password: xxxx
    external_labels:
     xxxx: "xxxxx"

scrape_configs:
- job_name: central-log
  loki_push_api:
    server:
      http_listen_port: 3500
      grpc_listen_port: 3600
    use_incoming_timestamp: true

**Expected behavior**
Promtail to start and accept logs from other promtail configurations on port 3500

**Environment:**
 - Infrastructure: Ubuntu 20.04.1 LTS
 - Deployment tool: docker compose

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.

Ah, I re-read your original description more closely:

> level=info ts=2022-09-28T07:15:52.266006699Z caller=pushtarget.go:85 msg="starting push server" job=central-log
panic: descriptor Desc{fqName: "promtail_central-log_tcp_connections", help: "Current number of accepted TCP connections.", constLabels: {}, variableLabels: [protocol]} is invalid: "promtail_central-log_tcp_connections" is not a valid metric name

This is quite unfortunate, but your line `job_name: central-log` is causing the issue. We use this job name to create metrics for each job, and because you have a hyphen the metric registration is panicking. See https://prometheus.io/docs/practices/naming/ for more details.

We could certainly do better here, though - for a start by sanitising the job name before we shove it into a metric.

Is this something you'd be interested in contributing? 
In any case, I've renamed this issue to accurately describe the problem and marked it as a bug.

Thank you for reporting!

--- a/clients/pkg/promtail/targets/lokipush/pushtargetmanager.go
+++ b/clients/pkg/promtail/targets/lokipush/pushtargetmanager.go
@@ -69,7 +69,7 @@ func validateJobName(scrapeConfigs []scrapeconfig.Config) error {
 		}
 		jobNames[cfg.JobName] = struct{}{}
 
-		scrapeConfigs[i].JobName = strutil.SanitizeLabelName(cfg.JobName);
+		scrapeConfigs[i].JobName = strutil.SanitizeLabelName(cfg.JobName)
 	}
 	return nil
 }

--- a/clients/pkg/promtail/targets/lokipush/pushtargetmanager.go
+++ b/clients/pkg/promtail/targets/lokipush/pushtargetmanager.go
@@ -3,11 +3,11 @@ package lokipush
 import (
 	"errors"
 	"fmt"
-	"strings"
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/util/strutil"
 
 	"github.com/grafana/loki/clients/pkg/logentry/stages"
 	"github.com/grafana/loki/clients/pkg/promtail/api"
@@ -69,7 +69,7 @@ func validateJobName(scrapeConfigs []scrapeconfig.Config) error {
 		}
 		jobNames[cfg.JobName] = struct{}{}
 
-		scrapeConfigs[i].JobName = strings.Replace(cfg.JobName, " ", "_", -1)
+		scrapeConfigs[i].JobName = strutil.SanitizeLabelName(cfg.JobName);
 	}
 	return nil
 }

--- a/clients/pkg/promtail/targets/lokipush/pushtargetmanager_test.go
+++ b/clients/pkg/promtail/targets/lokipush/pushtargetmanager_test.go
@@ -58,6 +58,19 @@ func Test_validateJobName(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "validate with special characters",
+			configs: []scrapeconfig.Config{
+				{
+					JobName: "job$1-2!3@4*job",
+					PushConfig: &scrapeconfig.PushTargetConfig{
+						Server: server.Config{},
+					},
+				},
+			},
+			wantErr: false,
+			expectedJob: "job_1_2_3_4_job",
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {

